Packs/Zerohack_XDR/Integrations/ZerohackXDR/ZerohackXDR.py

```diff
@@ -0,0 +1,331 @@
+    :type client: ``Client``
+
+    :param max_results: Maximum amount of results to fetch from the API per severity level.
+    :type max_results: ``int``
+
+    :param last_run: This parameter contains the details about last time this integration was run.
+    :type last_run: ``Dict[str, int]``
+
+    :param first_fetch_time: This parameter contains the time from when to fetch incidents details in case last run is not setup.
+    :type first_fetch_time: ``Optional[int]``
+
+    :param min_severity: This is minimum level of the severity you want to query the zerohack XDR for.
+    :type min_severity: ``int``
+
+    :return:
+        A tuple containing two elements:
+            next_run (``Dict[str, int]``): Contains the timestamp that will be
+                    used in ``last_run`` on the next fetch.
+            incidents (``List[dict]``): List of incidents that will be created in XSOAR.
+
+    :rtype: ``Tuple[Dict[str, int], List[dict]]``
+    """
+
+    # Get the last fetch time. If not provided then go for the first fetch time.
+    last_fetch = last_run.get('last_fetch', None)
+    if last_fetch is None:
+        last_fetch = first_fetch_time
+    else:
+        last_fetch = int(last_fetch)
+
+    last_incident_time = cast(int, last_fetch)
+    last_fetch_timestamp = str(datetime.fromtimestamp(last_fetch))
+    severity_levels = ZEROHACK_SEVERITIES[ZEROHACK_SEVERITIES.index(min_severity):]
+    severity_levels.sort()
+    # Initializating the incidents dictionary and setting the severity levels.
+    incidents: List[Dict[str, Any]] = []
+
+    for severity in severity_levels:
+        required_results = math.floor(max_results/len(severity_levels))
+        response = client.get_alerts(max_results = required_results, severity_level = severity, start_time = last_fetch_timestamp)
+        if response["message_type"] != "d_not_f":
+            for alert in response["data"]:
+                attack_time = datetime.strptime(alert.get('attack_timestamp', '0'), DATE_FORMAT)
+                incident_created_time = int((attack_time - datetime(1970, 1, 1)).total_seconds())
+                if incident_created_time > last_fetch:
+                    incident_name = "Zerohack XDR " + alert['ids_threat_class']
+                    incident = {
+                        'name': incident_name,
+                        'occurred': timestamp_to_datestring(incident_created_time),
+                        'type': alert['ids_threat_class'],
+                        'movement_type': alert['type_of_threat'],
+                        'platform': alert['platform'],
+                        'attacker_rep': alert['ip_rep'],
+                        'rawJSON': json.dumps(alert),
+                        'severity': convert_to_demisto_severity(alert.get('ids_threat_severity', 'Low')),
+                    }
+                    incidents.append(incident)
+                    last_incident_time = incident_created_time
```

Comment:
As you fetch incidents for every severity seperatly, a better implimentatuion would be to save the last_fetch_time for every severity. see this scenario-
We are fetching maximum of 4 incidents from severity level 1 and 2.
We fetch 2 incidents from level 1 with creation time - 10:30 , 10:35 our last_incident_time = 10:35
There is another incidents from level 1 at 10:37 but as we are limited to 2 incidents we didn't fetch it yet.
We fetch 2 incidents from level 2 with creation time - 10:30, 10:40 our last_incident_time = 10:40

In our next fetch we will start from 10:40 missing our level 1 incident that occurred at 10:37.

A suggesion for a fix could be to save our last_incident_time per severity in a dictionary or list. then comapre to our last_run object and update the times by severity.
Let me know if you need any help with the implimentation :)

Reply:
Your suggestions were insightful in improving the integration.